dawn/src/dawn/Compiler/DawnCompiler.cpp

```diff
@@ -170,184 +209,203 @@ std::unique_ptr<OptimizerContext> DawnCompiler::runOptimizer(std::shared_ptr<SIR
+    return CG->generateCode();
   } else if(options_->Backend == "c++-naive-ico") {
-    CG = std::make_unique<codegen::cxxnaiveico::CXXNaiveIcoCodeGen>(
-        optimizer->getStencilInstantiationMap(), *diagnostics_, options_->MaxHaloPoints);
+    auto CG = std::make_unique<codegen::cxxnaiveico::CXXNaiveIcoCodeGen>(
+        stencilInstantiationMap, *diagnostics_, options_->MaxHaloPoints);
+    return CG->generateCode();
   } else if(options_->Backend == "cuda") {
     const Array3i domain_size{options_->domain_size_i, options_->domain_size_j,
                               options_->domain_size_k};
-    CG = std::make_unique<codegen::cuda::CudaCodeGen>(
-        optimizer->getStencilInstantiationMap(), *diagnostics_, options_->MaxHaloPoints,
-        options_->nsms, options_->maxBlocksPerSM, domain_size);
+    auto CG = std::make_unique<codegen::cuda::CudaCodeGen>(stencilInstantiationMap, *diagnostics_,
+                                                           options_->MaxHaloPoints, options_->nsms,
+                                                           options_->maxBlocksPerSM, domain_size);
+    return CG->generateCode();
   } else if(options_->Backend == "c++-opt") {
     dawn_unreachable("GTClangOptCXX not supported yet");
   } else {
     diagnostics_->report(buildDiag("-backend", options_->Backend,
                                    "backend options must be : " +
                                        dawn::RangeToString(", ", "", "")(std::vector<std::string>{
                                            "gridtools", "c++-naive", "c++-opt", "c++-naive-ico"})));
-    return nullptr;
+    throw std::runtime_error("An error occurred.");
+  }
+}
+
+std::unique_ptr<codegen::TranslationUnit>
+DawnCompiler::compile(const std::shared_ptr<SIR>& stencilIR) {
+  diagnostics_->clear();
+  diagnostics_->setFilename(stencilIR->Filename);
+
+  IIRSerializer::Format serializationKind = IIRSerializer::Format::Json;
+  if(options_->SerializeIIR || (options_->DeserializeIIR != "")) {
+    if(options_->IIRFormat == "json") {
+      serializationKind = IIRSerializer::Format::Json;
+    } else if(options_->IIRFormat == "byte") {
+      serializationKind = IIRSerializer::Format::Byte;
+    } else {
+      dawn_unreachable("Unknown SIRFormat option");
+    }
+  }
+
+  // Check if options are valid
+  // -max-halo
+  if(options_->MaxHaloPoints < 0) {
+    diagnostics_->report(buildDiag("-max-halo", options_->MaxHaloPoints,
+                                   "maximum number of allowed halo points must be >= 0"));
+    throw std::runtime_error("An error occurred.");
+  }
+
+  std::map<std::string, std::shared_ptr<iir::StencilInstantiation>> stencilInstantiationMap;
+
+  if(options_->DeserializeIIR == "") {
+    stencilInstantiationMap = parallelize(stencilIR);
+    if(!options_->Debug) {
+      stencilInstantiationMap = optimize(stencilInstantiationMap);
+    }
+  } else {
+    throw std::runtime_error("Trying to deserialize IIR.");
```

Comment:
why is this throwing? The path should:
```
if input == SIR:
  IIR = parallelize(SIR)
else
  IIR = deserialize(input)
if not debug:
  IIR = optimize(IIR)
Code = generate(IIR)
```

Reply:
This was a leftover testing whether that code path is actually exercised. And actually all the tests continued to pass, so I'm going to take a hard line and leave it in until one breaks it ;)

Reply:
that is true since we don't have any deserializaton tests. But I would advise not merging something that clearly and purposely breaks functionality (even if it is not tested)

Reply:
Yes, fair point :) I took it out for now.